vision/dcgan_mnist/dcgan.jl

```diff
@@ -0,0 +1,130 @@
+using Base.Iterators: partition
+using Flux
+using Flux.Optimise: update!
+using Flux: logitbinarycrossentropy
+using Images
+using MLDatasets
+using Statistics
+using Parameters
+using Printf
+
+@with_kw struct HyperParams
+    batch_size::Int = 128
+    latent_dim::Int = 100
+    epochs::Int = 20
+    verbose_freq::Int = 1000
+    output_x::Int = 6
+    output_y::Int = 6
+    lr_dscr::Float64 = 0.0002
+    lr_gen::Float64 = 0.0002
+end
+
+function create_output_image(gen, fixed_noise, hparams)
+    @eval Flux.istraining() = false
+    fake_images = @. cpu(gen(fixed_noise))
+    @eval Flux.istraining() = true
+    image_array = dropdims(reduce(vcat, reduce.(hcat, partition(fake_images, hparams.output_y))); dims=(3, 4))
+    image_array = @. Gray(image_array + 1f0) / 2f0
+    return image_array
+end
+
+# Loss functions
+function discriminator_loss(real_output, fake_output)
+    real_loss = mean(logitbinarycrossentropy.(real_output, 1f0))
+    fake_loss = mean(logitbinarycrossentropy.(fake_output, 0f0))
+    return real_loss + fake_loss
+end
+
+generator_loss(fake_output) = mean(logitbinarycrossentropy.(fake_output, 1f0))
+
+function train_discriminator!(gen, dscr, batch, opt_dscr, hparams)
+    noise = randn(Float32, hparams.latent_dim, hparams.batch_size) |> gpu
+    fake_input = gen(noise)
+    ps = Flux.params(dscr)
+    # Taking gradient
+    loss, back = Flux.pullback(ps) do
+        discriminator_loss(dscr(batch), dscr(fake_input))
+    end
+    grad = back(1f0)
+    update!(opt_dscr, ps, grad)
+    return loss
+end
+
+function train_generator!(gen, dscr, batch, opt_gen, hparams)
+    noise = randn(Float32, hparams.latent_dim, hparams.batch_size) |> gpu
+    ps = Flux.params(gen)
+    # Taking gradient
+    loss, back = Flux.pullback(ps) do
+        generator_loss(dscr(gen(noise)))
+    end
+    grad = back(1f0)
+    update!(opt_gen, ps, grad)
+    return loss
+end
+
+function train()
+    # Model Parameters
+    hparams = HyperParams()
+
+    # Load MNIST dataset
+    images, _ = MLDatasets.MNIST.traindata(Float32)
+    # Normalize to [-1, 1] and convert it to WHCN
+    image_tensor = permutedims(reshape(@.(2f0 * images - 1f0), 28, 28, 1, :), (2, 1, 3, 4))
```

Comment:
ouch, the permutedims is a bit annoying but I guess we cannot do much about it without breaking changes in MLDatasets

Reply:
good catch anyways